app/src/main/kotlin/com/wire/android/ui/authentication/login/LoginScreen.kt

```diff
@@ -77,27 +87,68 @@ import kotlinx.coroutines.launch
 @Composable
 fun LoginScreen() {
     val loginViewModel: LoginViewModel = hiltViewModel()
+    val loginEmailViewModel: LoginEmailViewModel = hiltViewModel()
 
     LoginContent(
         onBackPressed = { loginViewModel.navigateBack() },
         loginViewModel,
+        loginEmailViewModel,
         ssoLoginResult = loginViewModel.ssoLoginResult
     )
 }
 
+@OptIn(ExperimentalAnimationApi::class)
+@Composable
+private fun LoginContent(
+    onBackPressed: () -> Unit,
+    viewModel: LoginViewModel,
+    loginEmailViewModel: LoginEmailViewModel,
+    ssoLoginResult: DeepLinkResult.SSOLogin?
+) {
+    val navController = rememberTrackingAnimatedNavController() { LoginNavigationItem.fromRoute(it)?.itemName }
+    Column(modifier = Modifier.fillMaxSize()) {
+        AnimatedNavHost(
+            navController = navController,
+            startDestination = LoginNavigationItem.MAIN_LOGIN_FORM_SELECTION.route
+        ) {
+            composable(
+                route = LoginNavigationItem.MAIN_LOGIN_FORM_SELECTION.route,
+                enterTransition = { smoothSlideInFromRight() },
+                exitTransition = { smoothSlideOutFromLeft() },
+                content = { MainLoginContent(onBackPressed, viewModel, loginEmailViewModel, ssoLoginResult) }
+            )
+            composable(
+                route = LoginNavigationItem.EMAIL_SECOND_FACTOR_INPUT.route,
+                enterTransition = { smoothSlideInFromRight() },
+                exitTransition = { smoothSlideOutFromLeft() },
+                content = { LoginEmailVerificationCodeScreen(loginEmailViewModel) }
+            )
+        }
+    }
+    val targetRoute: String = if (loginEmailViewModel.secondFactorVerificationCodeState.isCodeSent) {
+        LoginNavigationItem.EMAIL_SECOND_FACTOR_INPUT.route
+    } else {
+        LoginNavigationItem.MAIN_LOGIN_FORM_SELECTION.route
+    }
+    if (navController.currentDestination?.route != targetRoute) {
+        navController.navigate(targetRoute)
+    }
```

Comment:
@saleniuk @trOnk12 could you give me some input?

I feel completely stupid using navigation on Compose 😅

My attempt was making the Composable the one with the responsibility over selecting what should be shown. _i.e._ UI handles UI, ViewModel handles states and operations. If we change how 2FA code input is displayed, `LoginScreen` changes and the `ViewModel` shouldn't be affected in any way.

However, I don't feel confident in this approach as I haven't seen this exact approach in the app before. Perhaps there's a more common way of doing it?

Reply:
It's used in the app in a few places. For instance, `NewConversationRouter` passes the navigation lambda to the child composables which is responsible for navigating to another screen, the state or the whole common ViewModel can be passed as well. 
In `CreateTeamScreen` and `CreatePersonalAccountScreen` the common ViewModel has a Flow `var moveToStep = MutableSharedFlow` which is observed in the parent compose to navigate to the specific screen.
If you want something simpler, you can take a look at `SearchPeopleRouter`, where simple `Crossfade` is used to switch between two screens depending on the current state.

Reply:
> I feel that the logic and computation of the final route would make more sense if it lived in the VM. There we could check the appropriate flags and decide when to navigate to one screen or the other.

I'm a bit skeptical on this one.
Say we have a different design for bigger screens for example, that would display info side-by-side.
The ViewModel would need to change because it's the one calling navigation methods?

Or, in this case, if instead of navigating to a different screen, we just show a dialog or BottomSheet with the 2FA input.
The ViewModel would need to change because the presentation changed.

Instead of conveying state, the ViewModel is dictating presentation and making everything entangled.

The ViewModel could say `isLoginSuccessful = true`, but the ultimate entity to say where it should navigate to (_i.e._ what to present) should be the UI/presentation itself.


> It's used in the app in a few places. For instance, `NewConversationRouter` passes the navigation lambda to the child composables which is responsible for navigating to another screen, the state or the whole common ViewModel can be passed as well. In `CreateTeamScreen` and `CreatePersonalAccountScreen` the common ViewModel has a Flow `var moveToStep = MutableSharedFlow` which is observed in the parent compose to navigate to the specific screen. If you want something simpler, you can take a look at `SearchPeopleRouter`, where simple `Crossfade` is used to switch between two screens depending on the current state.

Awesome. Good to know this pattern isn't 100% new in the app :P I'll take a look at those.

Reply:
Yes, I totally agree that VM should definitely not dictate in detail how the UI behaves or navigates too in a reactive architectural pattern. However, our whole app and set of VM's is full of calls `navigationManager.navigate()` which is ultimately deciding the final navigation.
Therefore, maybe this owes a proper rediscussion on our current navigation patterns, otherwise I feel that if a change like you described comes up (redesign for larger screens), we will need to refactor anyways 80% of the VMs.